### ISSUE TITLE
Zod Preprocess returning unknown type

### ISSUE THREAD

**Issue by enchorb** [colinhacks/zod#3537]
The type returned from [z.preprocess](https://zod.dev/?id=preprocess) is giving out type "unknown"

Some Examples:

```
export const string_validation = z.preprocess(
  (value) => {
    if (typeof value !== 'string') return value;
    return value.trim().replace(/\s\s+/g, ' ');
  },
  z.string().min(1)
);

export const boolean_validation = z.preprocess((bool) => {
  if (typeof bool === 'boolean') return bool;
  return bool === 'true';
}, z.boolean());

export const number_validation = z.preprocess(
  (num) => (!num && num !== 0 ? num : Number(num)),
  z.number()
);
```

**Comment by samchungy** [comment]
This depends on what type util you're using to infer the type? Are you using infer, input or output?

**Comment by enchorb** [comment]
I use `infer` for my types and that works fine, types are as expected. This fails after `safeParse` / `parse` where the output for all preprocessed types is `unknown`


```
const product_validation = z.object({
   name: string_validation
   description: z.string(),
   hidden: boolean_validation.default(false),
   popular: z.boolean()
}).partial();

// Types Are As Expected
export type Product = z.infer<typeof product_validation>; 

// Types Are Unknown For All Preprocessed Types
const parsed = product_validation.safeParse({ name: 'Test Product', description: 'This is a product', hidden: false, popular: true )}.data; 
```

**Comment by samchungy** [comment]
Looks to work fine in this [playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbwF4F84DMoRHA5EiAE1wG4AoMgUwA9JY4BjCAOwGd52phmBzAfQBuAQwA2wQkJjAWcALxwkAOjBRKKiA0qtWACjJw4O4SICulAJRyAfIn0G4wdIZgBPMJQhPjZuAEJZ8ric3Dy4lqowJlDMcN6U5PZwEVExcYowXCA65oqqYCJCmjoA9AA6rOUA1MU8ADR4eOYJcCi1dkrBvNmKINw6AIzmZE0UNHTwTGzwAEYQECKUQsyCouKS0jHySipqWJraOjqz85ayNgh2js5uHk7HInIBePeLzGFJlJHRcPfNyd-3R6BDJmUhkVoKRQvJbZEZUWjQCYsdhwZgmEDTShQFZiCRSGRbZR5PZaXR2HRokCnGw6XyUuAAMgZqPRfieAAY4AB+Fk4ABccAAcujMVAKejzOY2gYlJTRdlhuQyJMUepCCYGDAcWt8ZtIRBpgArSianQXezMIQgSgCzr8Yw6jbSgyELQMLhgXUCjoZELZZ1wAAW4ldzAF0OWDrxG0UrvQQhMIhgOnjIlYFgDkDAiaEUG9ULmCxhQxQOTAuakomySuKxTgABUbqw4ABBVSt5sAUVoJpglEI8PGcFc7jgAAUsOrNXJIdx0FiADwj25wNUarVR9YsKwkOBkWuJQ9wAB6XLIQA)

<img width="476" alt="image" src="https://github.com/user-attachments/assets/a25363ef-5f68-48e2-a3fe-8ee9297f6de7">

**Comment by samchungy** [comment]
Looks fine to me in this [playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbwF4F84DMoRHA5EiAE1wG4AoMgUwA9JY4BjCAOwGd52phmBzAfQBuAQwA2wQkJjAWcALxwkAOjBRKKiA0qtWACjJw4O4SICulAJRyAfIn0G4wdIZgBPMJQhPjZuAEJZ8ric3Dy4lqowJlDMcN6U5PZwEVExcYowXCA65oqqYCJCmjoA9AA6rOUA1MU8ADR4eOYJcCi1dkrBvNmKINw6AIzmZE0UNHTwTGzwAEYQECKUQsyCouKS0jHySipqWJraOjqz85ayNgh2js5uHk7HInIBePeLzGFJlJHRcPfNyd-3R6BDJmUhkVoKRQvJbZEZUWjQCYsdhwZgmEDTShQFZiCRSGRbZR5PZaXR2HRokCnGw6XyUuAAMgZqPRfieAAY4AB+Fk4ABccAAcujMVAKejzOY2gYlJTRdlhuQyJMUepCCYGDAcWt8ZtIRBpgArSianQXezMIQgSgCzr8Yw6jbSgyELQMLhgXUCjoZELZZ1wAAW4ldzAF0OWDrxG0UrvQQhMIhgOnjIlYFgDkDAiaEUG9ULmCxhQxQOTAuakomySuKxTgABUbqw4ABBVSt5sAUVoJpglEI8PGcFc7jgAAUsOrNXJIdx0FiADwj25wNUarVR9YsKwkOBkWuJQ9wAB6XIoKvg6DmM7Xmu10ZYylz6bNpf3dfsp6AA)

<img width="503" alt="image" src="https://github.com/user-attachments/assets/75046c35-ba26-4a1b-8a97-9ab5d2e6271e">

**Comment by andrew-pledge-io** [comment]
I think I have a similar issue. For me it appears when using `z.ZodType`.

See this [playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbwF4F84DMoRHA5EiAE1wG4AoMgYwgDsBneMKAUyYkubroGVKALZiACGcALxwkAOiassHLgAoFANyEAbAJRiAfHBXqxo8blxwA-HBoBXNWrgAuOKs0AaCZOrMoHSYSExmBQ0Ncmp6eBoIAAUWNnkefkERcSlPb2Zff0CQinQrGkoYYFo4QghuIXRmKKEoOmYAHgAVbQU6ROFHKQAtIiaATzBGlq0EMjg4FhgrKBo4doFhSTpK6tr6hQQUDTIUCgB6fbh8gGtIgHc5gForuAAjLBPmGjMqWgY4MHXmQjFS8tWNTqgRkcU4CUWQg0mRgQnIhzgABEsnAAD7HGiEZjoYA0H5vcKfb6EABMf1BcnBvEhy0B3022xhcLICORATRGKxOLxhAJHy+wMIAGY-mUKlUgRtIjFZOwqR0oUz4Uc2cwOfkubj8WF+cSACx-aWxSlcalJWkS+lbaF+WEkIA).

```ts
import {z} from 'zod';

const preprocessSchema = z.preprocess((val) => (val === '' ? null : val), z.coerce.date());
const noPreprocessSchema = z.coerce.date();

function doSafeParse<T>(schema: z.ZodType<T>) {
  return schema.safeParse({})
}

// unknown -- broken?
const parsed = doSafeParse(preprocessSchema).data;
// Date | undefined
const parsed2 = preprocessSchema.safeParse({}).data;
// Date | undefined
const parsed3 = doSafeParse(noPreprocessSchema).data;
// Date | undefined
const parsed4 = noPreprocessSchema.safeParse({}).data;
```

**Comment by samchungy** [comment]
> I think I have a similar issue. For me it appears when using `z.ZodType`.
> 
> See this [playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbwF4F84DMoRHA5EiAE1wG4AoMgYwgDsBneMKAUyYkubroGVKALZiACGcALxwkAOiassHLgAoFANyEAbAJRiAfHBXqxo8blxwA-HBoBXNWrgAuOKs0AaCZOrMoHSYSExmBQ0Ncmp6eBoIAAUWNnkefkERcSlPb2Zff0CQinQrGkoYYFo4QghuIXRmKKEoOmYAHgAVbQU6ROFHKQAtIiaATzBGlq0EMjg4FhgrKBo4doFhSTpK6tr6hQQUDTIUCgB6fbh8gGtIgHc5gForuAAjLBPmGjMqWgY4MHXmQjFS8tWNTqgRkcU4CUWQg0mRgQnIhzgABEsnAAD7HGiEZjoYA0H5vcKfb6EABMf1BcnBvEhy0B3022xhcLICORATRGKxOLxhAJHy+wMIAGY-mUKlUgRtIjFZOwqR0oUz4Uc2cwOfkubj8WF+cSACx-aWxSlcalJWkS+lbaF+WEkIA).
> 
> ```ts
> import {z} from 'zod';
> 
> const preprocessSchema = z.preprocess((val) => (val === '' ? null : val), z.coerce.date());
> const noPreprocessSchema = z.coerce.date();
> 
> function doSafeParse<T>(schema: z.ZodType<T>) {
>   return schema.safeParse({})
> }
> 
> // unknown -- broken?
> const parsed = doSafeParse(preprocessSchema).data;
> // Date | undefined
> const parsed2 = preprocessSchema.safeParse({}).data;
> // Date | undefined
> const parsed3 = doSafeParse(noPreprocessSchema).data;
> // Date | undefined
> const parsed4 = noPreprocessSchema.safeParse({}).data;
> ```

You aren't declaring your function types correctly. Please read [this](https://github.com/colinhacks/zod#writing-generic-functions).

**Comment by cewald** [comment]
I noticed, that I get the right types if I set `compilerOptions.strict` to false – but not the other way around.

**Comment by comxd** [comment]
Since is not fixed you can force the return type like this:

```
export const trimed = z.preprocess(
  (value) => {
    return value.trim();
  },
  z.string().min(1)
) as ZodEffects<ZodString, string, string>;
```

**Comment by vagnerlandio** [comment]
@spoptchev  Same here, I also have problem with the handleSubmit:

```
Argument of type '(data: ZodFormData) => void' is not assignable to parameter of type 'SubmitHandler<TFieldValues>'.
  Types of parameters 'data' and 'data' are incompatible.
    Type 'TFieldValues' is not assignable to type '{ phoneNumber: string; }'.
      Property 'phoneNumber' is missing in type 'FieldValues' but required in type '{ phoneNumber: string; }'.ts(2345)
```

@comxd solved my problem, but I needed to add `unknown` before ZodEffects:
```
as unknown as ZodEffects<ZodString, string, string>
```

**Comment by paschaldev** [comment]
+1

I had this problem with react hook form and zodResolver

**Comment by uap-dev** [comment]
I think maybe some of the internal types have changed. 

My situation

```ts
export const coerceToISOString = (value: unknown): string => {
  if (value instanceof Date) {
    return value.toISOString()
  } else {
    return String(value)
  }
}
```

```ts
/// ...rest of zod object
endEffectiveDate: z.preprocess(
      coerceToISOString,
      z.iso.datetime('End Date is Required'),
    ) as z.ZodPipe<z.ZodTransform<string | Date, string>, z.ZodISODateTime>,
```

**Comment by masoudmoghaddari** [comment]
Same unknown type issue with react-hook-form

**Comment by Svish** [comment]
> I managed to fix this by overriding the type definition:
> 
> -- @ayloncarrijo

This fix only seem to work for v3. Anyone been able to fix this for v4?

Really wish this could be fixed somehow as it's rather frustrating not being able to use `z.preprocess` with `react-hook-form` and similar libraries... 😕

**Comment by Svish** [comment]
I've seen someone argue that with `z.preprocess` you don't know what the input type is because it could be anything, but that doesn't really make sense because then the input type of _all_ schemas really should be `unknown`. Even if you have `z.string()`, you don't know what will be passed to it, that's the point of the validation/parsing to begin with. The input type shouldn't be "what do we know it is", but "what do we _expect_ it to be".

If I have `z.preprocess((val) => ..., z.string())`, it should be irrelevant what the preprocessing does or doesn't do, the expected input type is obviously `string`.

A slightly sketchy "workaround" seem to be to declare the parameter type on the preprocess function, e.g. like this:
```ts
const accountNumberSchema = z.preprocess(
  (value: string | null | undefined) => typeof value === 'string'
    ? value.replace(/\D+/g, '')
    : value,
  z.stringFormat('accountNumber', isAccountNumber)
)

type Input = z.input<typeof accountNumberSchema>;
//   ^^ string | null | undefined
```
Sketchy, because we of course have no idea if what will pass through the preprocess function actually _is_ a string or not, so we need to remember to guard ourselves against that, even though typescript will not remind us.

The absolute best solution here would be if `z.preprocess` just "copied" the input type of whatever schema was passed to it as the second parameter.

Basically, the following two schemas really should have the exact same inferred input and output types, while the preprocessor function itself should still be `unknown`:
```ts
const test1 = z.string();
const test2 = z.preprocess((value) => value, z.string());
//                         ^ fn should always be `(value: unknown): unknown`
```

**Comment by bloodykheeng** [comment]
i think a easier approach is to  specify the input type generic
ie number<string>()

see code below
```
quantity: z.coerce.number<number>().min(1, "Quantity must be at least 1"),
total_price: z.coerce.number<number>().nullish()
```

if its not specified react hook form will try to infer and see unknown because of coerce but it the type generic is specified the 
it will see number

**Comment by Svish** [comment]
Easier if you're using `coerce`, but the issue here is with `preprocess` which has more complicated generics and is less straight forward to type correctly. Specifying generics directly in schemas also tend to make for brittle and less trustworthy types.

**Comment by bloodykheeng** [comment]
> Easier if you're using `coerce`, but the issue here is with `preprocess` which has more complicated generics and is less straight forward to type correctly. Specifying generics directly in schemas also tend to make for brittle and less trustworthy types.

i think this z preprocess has serious bugs in it huh
see here when i put the static type of the argument it takes them as the return type so it stops showing the unknown and when i remove them i get the unkown issue back
```
 (val: number | null | undefined)
```

here is it 
```
price: z.preprocess(
          (val: number | null | undefined) => {
            if (val === null || val === undefined) return null;
            const num = Number(val);
            return isNaN(num) ? null : num;
          },
          z.coerce.number<number>().nullish()
        ).optional()
```

my goal want to get a z.preprocess eqivalent of this
```
z.coerce.number<number>().nullish()
```

though it worked that z.prprocess but i think somethong needs to be resolved

**Comment by arthurvanl** [comment]
Had the same issue. However @comxd his example did not work for me.

Fixed it by doing this:

```ts
import { DateTime } from "luxon"
import z from "zod"

const dateInner = z.iso.datetime({ offset: true }).transform((val) => DateTime.fromISO(val));

export const dateSchemaV2 = z.preprocess(
    (val) => DateTime.isDateTime(val) ? val.toISO() : val,
    dateInner
) as unknown as z.ZodPipe<z.ZodTransform<string, DateTime | string>, typeof dateInner>;
```

**Comment by colinhacks** [comment]
Closing this out — the input type of `z.preprocess` is intentionally `unknown`. The whole point of preprocess is to massage arbitrary input before validation, so claiming a narrower input type would be unsound: there's no runtime guarantee about what gets passed in.

If you want a narrower input type for ergonomics (e.g. so `react-hook-form`'s `zodResolver` sees the right field type), annotate the preprocess function's parameter and `B` will be inferred from it:

```ts
const schema = z.preprocess(
  (val: string | null | undefined) => (val ?? "").trim(),
  z.string()
);
// z.input<typeof schema>  -> string | null | undefined
// z.output<typeof schema> -> string
```

You're on the hook for making sure the function actually handles every value of that declared type — TypeScript won't catch a mismatch between the annotation and reality — but as an opt-in escape hatch it does the job.
